### PR TITLE
Combine implementations of Global and Local

### PR DESCRIPTION
**Pull request by btracey**

### CODE REVIEW

**Comment by btracey** [comment]
This is step two of the plan. After this is comments and a real global method.

File: global.go [hunk 1]
```diff
@@ -209,119 +187,46 @@ func globalWorker(task int, m GlobalMethod, g *globalStatus, loc *Location, x []
...
-	if s != NotTerminated {
-		return s
+	if status != NotTerminated {
+		return status
 	}
 	switch op {
 	case NoOperation:
 	case InitIteration:
-		panic("optimize: GlobalMethod return InitIteration")
+		panic("optimize: Method return InitIteration")
```
Comment (vladimir-ch) [inline global.go]: `... returned ...`

File: guessandcheck.go [hunk 2]
```diff
@@ -22,14 +31,27 @@ func (g *GuessAndCheck) Done() {
 	// No cleanup needed
 }
 
-func (g *GuessAndCheck) InitGlobal(tasks int) int {
+func (g *GuessAndCheck) InitGlobal(dim, tasks int) int {
 	g.eval = make([]bool, tasks)
+	g.bestF = math.Inf(1)
+	g.bestX = make([]float64, dim)
```
Comment (vladimir-ch) [inline guessandcheck.go]: Reuse `g.bestX` if possible?

File: minimize.go [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+
+// checkConvergence returns NotTerminated if the Location does not satisfy the
+// convergence criteria given by settings. Otherwise a corresponding status is
+// returned.
+// Unlike checkLimits, checkConvergence is called only at MajorIterations.
+//
+// If local is true, gradient convergence is also checked.
+func checkConvergence(loc *Location, settings *Settings, local bool) Status {
+	if local {
+		if loc.Gradient != nil {
```
Comment (vladimir-ch) [inline minimize.go]: Merge this with `if local` ?

File: minimize.go [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+	if op&GradEvaluation != 0 {
+		stats.GradEvaluations++
+	}
+	if op&HessEvaluation != 0 {
+		stats.HessEvaluations++
+	}
+}
+
+// checkLimits returns NotTerminated status if the various limits given by
+// settings has not been reached. Otherwise it returns a corresponding status.
```
Comment (vladimir-ch) [inline minimize.go]: s/has/have/

File: minimize.go [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+// Unlike checkConvergence, checkLimits is called by Local at _every_ iteration.
```
Comment (vladimir-ch) [inline minimize.go]: ... is called by Local and Global at every iteration.

File: global.go (hunk 1 above)
Comment (btracey) [inline global.go]: Done.

File: guessandcheck.go (hunk 2 above)
Comment (btracey) [inline guessandcheck.go]: done.

File: minimize.go (hunk 4 above)
Comment (btracey) [inline minimize.go]: Done.

**Pull request merged**